crates/oxc_linter/src/rules/react/style_prop_object.rs

```diff
@@ -0,0 +1,428 @@
+    /// Require that the value of the prop style be an object or a variable that is an object.
+    ///
+    /// ### Examples
+    ///
+    /// Examples of **incorrect** code for this rule:
+    /// ```jsx
+    /// <div style="color: 'red'" />
+    /// <div style={true} />
+    /// <Hello style={true} />
+    /// const styles = true;
+    /// <div style={styles} />
+    ///
+    /// React.createElement("div", { style: "color: 'red'" });
+    /// React.createElement("div", { style: true });
+    /// React.createElement("Hello", { style: true });
+    /// const styles = true;
+    /// React.createElement("div", { style: styles });
+    /// ```
+    ///
+    /// Examples of **correct** code for this rule:
+    /// ```jsx
+    /// <div style={{ color: "red" }} />
+    /// <Hello style={{ color: "red" }} />
+    /// const styles = { color: "red" };
+    /// <div style={styles} />
+    ///
+    /// React.createElement("div", { style: { color: 'red' }});
+    /// React.createElement("Hello", { style: { color: 'red' }});
+    /// const styles = { height: '100px' };
+    /// React.createElement("div", { style: styles });
+    /// ```
+    StylePropObject,
+    suspicious
+);
+
+fn is_invalid_expression(expression: Option<&Expression>, ctx: &LintContext<'_>) -> bool {
+    let Some(expression) = expression else {
+        return false;
+    };
+
+    match expression {
+        Expression::StringLiteral(_) | Expression::BooleanLiteral(_) => true,
+        Expression::Identifier(ident) => {
+            let Some(node) = get_declaration_of_variable(ident, ctx) else {
+                return false;
+            };
+
+            let AstKind::VariableDeclarator(var_decl) = node.kind() else {
+                return false;
+            };
+
+            return is_invalid_expression(var_decl.init.as_ref(), ctx);
+        }
+        _ => false,
+    }
+}
+
+fn is_invalid_style_attribute(attribute: &JSXAttribute, ctx: &LintContext<'_>) -> bool {
+    let JSXAttributeName::Identifier(attr_ident) = &attribute.name else {
+        return false;
+    };
+
+    if attr_ident.name == "style" {
+        if let Some(attr_value) = &attribute.value {
+            return match attr_value {
+                JSXAttributeValue::StringLiteral(_) => true,
+                JSXAttributeValue::ExpressionContainer(container) => {
+                    return is_invalid_expression(container.expression.as_expression(), ctx);
+                }
+                _ => false,
+            };
+        }
+    }
+
+    false
+}
+
+impl Rule for StylePropObject {
+    fn from_configuration(value: serde_json::Value) -> Self {
+        Self(Box::new(StylePropObjectConfig {
+            allow: value
+                .get(0)
+                .and_then(|v| v.get("allow"))
+                .and_then(serde_json::Value::as_array)
+                .map(|v| {
+                    v.iter()
+                        .filter_map(serde_json::Value::as_str)
+                        .map(ToString::to_string)
```

Comment:
use `CompactStr` instead of `String`s

Reply:
Also, if you sort this, you can use a binary search for `O(log(n))` containment checks (instead of `.contains`, which is `O(n)`.